docs/rfc/rfc-002-wasm-contract-parallelization.md

```diff
@@ -0,0 +1,37 @@
+
+> OPEN QUESTION: If we require additional granularity, should we allow mapping specific to the contract instance (this would be relevant for other contracts calling a specific contract (eg. code ID 1), if there are multiple instances of thcontracts with code ID 1, we could allow them to run in parallel since the other calling contract can specify exactly what resources would be affected that are specific to the contract instance that it would be calling).
+
+This would follow a similar pattern for the message dependency mapping for sdk messages, but instead would be blocked prior to contract execution and would release the resources at the end of the contract execution. The reason for this is because it would be much more effort to construct a system to granularly release resources within contract execution.
+
+One key difference would be that we would also store an enabled flag for the parallelization mapping, and if the wasm contract fails validation, we would disable the parallelization mappings and only process that contract sequentially until the parallellization mappings are updated.
+
+
+### Failure Remediation
+However, we do need some remediation steps for wasm contract execution as well in the case of improper resource access that hasn't been defined in the dependency mappings. One way we can accomplish this is by modifying the wasm module to perform ongoing resource validation during a contract's execution. Because wasm contract need to go through the wasm module for any queries or messages during contract execution, we can either fork the wasm module or override the message handler and querier plugins so that we can have greater control over all of the queries / messages that are executed by the wasm contract. As we process these messages / queries, we can validate them against the wasm contract mappings, and if we detect a resource access that has not been defined in the contract parallelization, we can then fail the contract's transaction, and disable the parallel processing via the enabled flag for that contract.
+
+We need to fail the current transaction so that it doesn't improperly modify any resources that haven't been properly defined in the contract mapping. Additionally, we would also need to fail any other transactions with that contract in the current block, since otherwise we would need to rebuild the dag with the sequential processing for the contract and restart the block's execution. Then, in the following blocks, those contract transactions would be able to process sequentially as expected.
+
+> OPEN QUESTION: What are the potential side effects of failing contract transactions this way?
+>
+> OPEN QUESTION: What about the specific transactions that failed? Is it ok to just leave them as failed and assume that the user can resend them later? Or do we need to some other handling to ensure that we process those TXs sequentially as well (This would negatively impact block time)?
```

Comment:
i think we can get away with just leaving them as failed since technically it's on contract developers who specified the dependency wrong. How strong this argument is though depends on how easy we make dependency registration

Reply:
The behavior would only fail on that singular block, since the registration would then be disabled and future contract executions would be processed sequentially, so its not super consequential.